### PR TITLE
DiCE tests

### PR DESCRIPTION
**Pull request by raam93**
adding DiCE test cases

### CODE REVIEW

File: tests/test_model_interface/test_keras_tensorflow_model.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    def test_model_output(self):
+        self.m.load_model()
+        test_instance = np.array([[0.5]*29], dtype=np.float32)
+        if tf.__version__[0] == '1':
+            input_instance = tf.Variable(test_instance, dtype=tf.float32)
+            output_instance = self.m.get_output(input_instance)
+            prediction = self.sess.run(output_instance, feed_dict={input_instance:test_instance})[0][0]
+        else:
+            prediction = self.m.get_output(test_instance).numpy()[0][0]
+        assert (round(prediction,4)-0.747)<1e-6
```
Comment (amit-sharma) [inline tests/test_model_interface/test_keras_tensorflow_model.py:52]: a comment on how these numbers are computed will be useful for someone to double-check later.

File: tests/test_model_interface/test_pytorch_model.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    def test_load_model(self):
+        self.m.load_model()
+        assert self.m.model is not None
+
+    def test_model_output(self):
+        self.m.load_model()
+        test_instance = pyt.tensor(np.array([[0.5]*29], dtype=np.float32))
+        prediction = self.m.get_output(test_instance).detach().numpy()[0][0]
+        assert (round(prediction,4)-0.0957)<1e-6
```
Comment (amit-sharma) [inline tests/test_model_interface/test_pytorch_model.py:40]: again, a comment on how the number is computed will be useful.

File: tests/test_data_interface/test_private_data_interface.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                    type_and_precision={'hours_per_week':['float', 2]}, mad={'age': 10})
+
+class TestPrivateDataMethods:
+    @pytest.fixture(autouse=True)
+    def _get_data_object(self, data_object):
+        self.d = data_object
+
+    def test_mads(self):
+        mads = self.d.get_valid_mads(normalized=False)
+        assert list(mads.values()) == [10.0, 1.0]
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_data_interface/test_private_data_interface.py]: yes it is

File: tests/test_data_interface/test_public_data_interface.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+def data_object():
+    dataset = helpers.load_adult_income_dataset()
+    return dice_ml.Data(dataframe=dataset, continuous_features=['age', 'hours_per_week'], outcome_name='income', permitted_range={'age':[45,60]}, continuous_features_precision={'hours_per_week': 2})
+
+class TestPublicDataMethods:
+    @pytest.fixture(autouse=True)
+    def _get_data_object(self, data_object):
+        self.d = data_object
+
+    def test_permitted_range(self):
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_data_interface/test_public_data_interface.py:16]: so permitted_range does things in a loop, so I think it's sufficient to test any one feature.

File: tests/test_dice.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import pytest
+
+import dice_ml
+from dice_ml.utils import helpers
+
+class TestBaseExplainerLoader:
+    def _get_exp(self, backend):
+        dataset = helpers.load_adult_income_dataset()
+        d = dice_ml.Data(dataframe=dataset, continuous_features=['age', 'hours_per_week'], outcome_name='income')
+        ML_modelpath = helpers.get_adult_income_modelpath(backend=backend)
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_dice.py:10]: just used it as a short form for Machine Learning

File: tests/test_dice_interface/test_dice_pytorch.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        query_instance = self.exp.data_interface.prepare_query_instance(query_instance=query_instance, encode=True)
+        self.query_instance = query_instance.iloc[0].values
+        self.exp.initialize_CFs(self.query_instance, init_near_query_instance=True)
+        self.exp.target_cf_class = torch.tensor(1).float()
+        np.random.seed(42)
+        weights = np.random.rand(len(self.exp.data_interface.encoded_feature_names))
+        self.exp.feature_weights_list = torch.tensor(weights)
+
+    @pytest.mark.parametrize("yloss, output",[("hinge_loss", "4.7753"), ("l2_loss", "0.9999"), ("log_loss", "4.2892")])
+    def test_first_loss(self, yloss, output):
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_dice_interface/test_dice_pytorch.py]: updated

File: tests/test_dice_interface/test_dice_pytorch.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        self.exp.yloss_type = yloss
+        loss1 = self.exp.compute_first_part_of_loss()
+        assert str(np.round(loss1.data.detach().numpy(),4)) == output
+
+    def test_second_loss(self):
+        self.exp.x1 =  torch.tensor(self.query_instance)
+        loss2 = self.exp.compute_second_part_of_loss()
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_dice_interface/test_dice_pytorch.py]: updated

File: tests/test_dice_interface/test_dice_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        assert str(np.round(loss2.data.detach().numpy(),4)) == "0.0068"
+
+    @pytest.mark.parametrize("diversity_loss, output",[("dpp_style:inverse_dist", "0.0104"), ("avg_dist", "0.1743")])
+    def test_third_loss(self, diversity_loss, output):
+        self.exp.diversity_loss_type = diversity_loss
+        loss3 = self.exp.compute_third_part_of_loss()
+        assert str(np.round(loss3.data.detach().numpy(),4)) == output
+
+    def test_fourth_loss(self):
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_dice_interface/test_dice_pytorch.py]: regularization loss

File: tests/test_dice_interface/test_dice_pytorch.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        loss4 = self.exp.compute_fourth_part_of_loss()
+        assert str(np.round(loss4.data.detach().numpy(),4)) == "0.2086"
+
+    def test_final_cfs_and_preds(self, query_instance):
+        dice_exp = self.exp.generate_counterfactuals(query_instance, total_CFs=4, desired_class="opposite")
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_dice_interface/test_dice_pytorch.py:62]: we are testing with a sample query instance from conftest.py

File: tests/test_dice_interface/test_dice_pytorch.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
+import pytest
+
+import dice_ml
+from dice_ml.utils import helpers
+import numpy as np
+
+torch = pytest.importorskip("torch")
+
+@pytest.fixture
+def pyt_exp_object():
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_dice_interface/test_dice_pytorch.py:10]: It is not straightforward to put pyt_exp_object in conftest.py or have a different scope for two reasons:
1. It is not trivial to pass parameters from one fixture to a parent one (https://github.com/pytest-dev/pytest/issues/1694), and both `_initiate_exp_object` and `pyt_exp_object` are fixtures.
2.  `_initiate_exp_object` needs to be a fixture and had to use it with `autouse=True` to avoid calling the fixture and initiating the explainer object repeatedly.
Perhaps, there's a way to address this, so I'm open to suggestions.

File: tests/test_dice_interface/test_dice_tensorflow.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            loss2 = self.exp.dice_sess.run(loss2, feed_dict={self.exp.x1: self.query_instance})
+        else:
+            self.exp.x1 = tf.constant(self.query_instance, dtype=tf.float32)
+            loss2 = self.exp.compute_second_part_of_loss()
+        assert str(np.round(loss2, 4)) == '0.0068'
+
+    @pytest.mark.parametrize("diversity_loss, output",[("dpp_style:inverse_dist", "0.0104"), ("avg_dist", "0.1743")])
+    def test_third_loss(self, diversity_loss, output):
+        if tf.__version__[0] == '1':
+            loss3 = self.exp.compute_third_part_of_loss(diversity_loss)
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_dice_interface/test_dice_tensorflow.py]: yes correct

File: tests/test_dice_interface/test_dice_pytorch.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    backend = 'PYT'
+    dataset = helpers.load_adult_income_dataset()
+    d = dice_ml.Data(dataframe=dataset, continuous_features=['age', 'hours_per_week'], outcome_name='income')
+    ML_modelpath = helpers.get_adult_income_modelpath(backend=backend)
+    m = dice_ml.Model(model_path= ML_modelpath, backend=backend)
+    exp = dice_ml.Dice(d, m)
+    return exp
+
+@pytest.fixture
```
[the remark this answers is not in the document]
Comment (raam93) [reply tests/test_dice_interface/test_dice_pytorch.py]: same reasoning for having `pyt_exp_object` fixture in `test_dice_pytorch.py`
